### PR TITLE
Add the transformop parameter in TensorReduceFunctorImpl

### PR DESCRIPTION
**Pull request by AnnaTrainingG**
### PR types

Others 
### PR changes

OPs
### Describe

In the old TensorReduceFunctorImpl API,developer cannot initialize transform. Now, users can pass the initialized transform into TensorReduceFunctorImpl as a parameter

### CODE REVIEW

**Comment by xingfeng01** [comment]
『users can pass the initialized transform into TensorReduceFunctorImpl as a parameter』在 python 层暴露吗 ？

**Comment by AnnaTrainingG** [comment]
修改的是C++接口， python层面没有进行修改，在 python 层不暴露

File: paddle/fluid/operators/reduce_ops/reduce_op.cu.h [hunk 1]
```diff
@@ -722,12 +722,12 @@ __global__ void ReduceHigherDimKernel(const Tx* x, Ty* y, ReduceOp reducer,
   }
 }
 
-template <typename Tx, typename Ty, typename MPType, typename ReduceOp>
+template <typename Tx, typename Ty, typename MPType, typename ReduceOp,
+          typename TransformOp>
 static void LaunchReduceKernel(const Tx* x_data, Ty* y_data,
-                               const ReduceOp& reducer, MPType init,
+                               const ReduceOp& reducer,
+                               const TransformOp transform, MPType init,
```
Comment (ZzSean) [inline paddle/fluid/operators/reduce_ops/reduce_op.cu.h]: reducer和transform是不是可以统一都传引用

File: paddle/fluid/operators/reduce_ops/reduce_prod_op.cu [hunk 2]
```diff
@@ -12,12 +12,12 @@
...
 REGISTER_OP_CUDA_KERNEL(
-    reduce_prod, ops::ReduceCudaKernel<float, paddle::operators::CustomMul>,
-    ops::ReduceCudaKernel<int, paddle::operators::CustomMul>,
-    ops::ReduceCudaKernel<double, paddle::operators::CustomMul>,
-    ops::ReduceCudaKernel<int64_t, paddle::operators::CustomMul>);
+    reduce_prod,
+    ops::ReduceCudaKernel<float, kps::AddFunctor, kps::IdentityFunctor>,
+    ops::ReduceCudaKernel<int, kps::AddFunctor, kps::IdentityFunctor>,
+    ops::ReduceCudaKernel<double, kps::AddFunctor, kps::IdentityFunctor>,
+    ops::ReduceCudaKernel<int64_t, kps::AddFunctor, kps::IdentityFunctor>);
```
Comment (ZzSean) [inline paddle/fluid/operators/reduce_ops/reduce_prod_op.cu]: prod的reducer是不是要传MulFunctor

File: paddle/fluid/operators/reduce_ops/reduce_op.cu.h (hunk 1 above)
Comment (AnnaTrainingG) [reply paddle/fluid/operators/reduce_ops/reduce_op.cu.h]: 已经修改

File: paddle/fluid/operators/reduce_ops/reduce_prod_op.cu (hunk 2 above)
Comment (AnnaTrainingG) [reply paddle/fluid/operators/reduce_ops/reduce_prod_op.cu]: 已经修改

**Review by zhwesky2010 — APPROVED** [review]
LGTM for triangular_solve op

**Review by zhiqiu — APPROVED** [review]
LGTM

**Comment by limin2021** [comment]
LGTM for modifications in attn_gemm.h

**Review by limin2021 — APPROVED** [review]

**Review by GuoxiaWang — APPROVED** [review]
LGTM for margin_cross_entropy

**Review by jeff41404 — APPROVED** [review]
LGTM for trace op

**Review by Liu-xiandong — APPROVED** [review]

**Review by xingfeng01 — APPROVED** [review]

**Pull request merged**